grpc/v1/webserver.proto

```diff
@@ -0,0 +1,625 @@
+  // EXPERIMENTAL: the behavior of this flag may be changed in future versions.
+  bool chunk_matches = 10;
+
+  // EXPERIMENTAL. If true, document ranks are used as additional input for
+  // sorting matches.
+  bool use_document_ranks = 11;
+
+  // EXPERIMENTAL. When UseDocumentRanks is enabled, this can be optionally set to adjust
+  // their weight in the file match score. If the value is <= 0.0, the default weight value
+  // will be used. This option is temporary and is only exposed for testing/ tuning purposes.
+  double document_ranks_weight = 12;
+
+  // Trace turns on opentracing for this request if true and if the Jaeger address was provided as
+  // a command-line flag
+  bool trace = 13;
+
+  // If set, the search results will contain debug information for scoring.
+  bool debug_score = 14;
+
+  // SpanContext is the opentracing span context, if it exists, from the zoekt client
+  map<string, string> span_context = 15;
+}
+
+message Q {
+  oneof query {
+    RawConfig raw_config = 1;
+    Regexp regexp = 2;
+    Symbol symbol = 3;
+    Language language = 4;
+    Const const = 5;
+    Repo repo = 6;
+    RepoRegexp repo_regexp = 7;
+    BranchesRepos branches_repos = 8;
+    RepoIds repo_ids = 9;
+    RepoSet repo_set = 10;
+    FileNameSet file_name_set = 11;
+    Type type = 12;
+    Substring substring = 13;
+    And and = 14;
+    Or or = 15;
+    Not not = 16;
+    Branch branch = 17;
+  }
+}
+
+// RawConfig filters repositories based on their encoded RawConfig map.
+message RawConfig {
+  enum Flags {
+    UNKNOWN = 0x00;
+    ONLY_PUBLIC = 0x01;
+    ONLY_PRIVATE = 0x02;
+    ONLY_FORKS = 0x04;
+    NO_FORKS = 0x08;
+    ONLY_ARCHIVED = 0x10;
+    NO_ARCHIVED = 0x20;
+  }
+
+  uint64 flags = 1;
+}
+
+// Regexp is a query looking for regular expressions matches.
+message Regexp {
+  string regexp = 1;
+  bool file_name = 2;
+  bool content = 3;
+  bool case_sensitive = 4;
+}
+
+message Symbol {
+  Q expr = 1;
+}
+
+message Language {
+  string language = 1;
+}
+
+message Const {
+  bool value = 1;
+}
+
+// TODO: what's the difference between Repo and RepoRegexp?
+message Repo {
+  string regexp = 1;
+}
+
+// TODO: what's the difference between Repo and RepoRegexp?
+message RepoRegexp {
+  string regexp = 1;
+}
+
+// BranchesRepos is a slice of BranchRepos to match.
+message BranchesRepos {
+  repeated BranchRepos list = 1;
+}
+
+// BranchRepos is a (branch, sourcegraph repo ids bitmap) tuple. It is a
+// Sourcegraph addition.
+message BranchRepos {
+  string branch = 1;
+  // a serialized roaring bitmap of the target repo ids
+  bytes repos = 2;
+}
+
+// Similar to BranchRepos but will be used to match only by repoid and
+// therefore matches all branches
+message RepoIds {
+  // a serialized roaring bitmap of the target repo ids
+  bytes repos = 1;
+}
+
+// RepoSet is a list of repos to match.
+message RepoSet {
+  map<string, bool> set = 1;
```

Comment:
do we have a set in protobuf? This is only a map in go because there is no set.

Reply:
> This is only a map in go because there is no set.

Why is it `map[string]bool` rather than `map[string]struct{}` then?

Reply:
> do we have a set in protobuf?

No, it doesn't. It would be represented on the wire the same as `repeated T`. In fact, `map<string, bool>` is just syntax sugar for `repeated { string key = 1; bool value = 2; }`

Reply:
Ok. It was a map[string]bool due to me being lazy and liking the look of `if set[k]` more than `if _, ok := set[k]; ok`. I incorrectly thought back then it took up the same amount of space in memory between the two alternatives.

Reply:
hahaha, okay, cool. That makes sense 😄